src/request.js

```diff
@@ -0,0 +1,237 @@
+let http = require('http');
+let https = require('https');
+let assert = require('assert');
+let urllib = require('url');
+let debug = require('debug')('requests');
+
+let versionString = '/unknownversion';
+
+// Figure out the version string that we'll include in the user agent.  If we
+// cannot figure out the version string, we'll (in the future) figure out what
+// git/hg revision we're on and use that.
+try {
+  let pkgInfo = require('../package.json');
+  versionString = '/' + pkgInfo.version;
+} catch (err) { }
+
+let httpAgent = new http.Agent({
+  keepAlive: true,
+});
+
+let httpsAgent = new https.Agent({
+  keepAlive: true,
+});
+
+/**
+ * Return true if these headers are ones that we'll support and false if they
+ * aren't.  We only support headers where there is a single level deep mapping
+ * of string to string values
+ */
+function validateHeaders(headers) {
+  if (typeof headers !== 'object') {
+    return false;
+  }
+  for (let key of Object.keys(headers)) {
+    if (typeof headers[key] !== 'string') {
+      return false;
+    }
+  }
+  return true;
+}
+
+class RequestError extends Error {
+
+}
+
+/**
+ * Make a request against a URL.  The provided URL will be a correctly formed
+ * URL in a string.  This method will not follow any redirects.  This method
+ * will return a promise which resolves an object like {response, request}.
+ */
+async function request(url, opts = {}) {
+  assert(typeof url === 'string', 'must provide a url');
+  let urlParts = urllib.parse(url);
```

Comment:
exploding `urlParts` might make the code simpler... `let {protocol, port, ...} = URL.parse(url);`